README.md
# flutter_image_utils

[![pub package](https://img.shields.io/pub/v/flutter_image_utils.svg?style=flat-square)](https://pub.dartlang.org/packages/flutter_image_utils)

Crops image as native plugin, both iOS and Android

## Usage

[Full example](https://github.com/vanelizarov/flutter_image_utils/blob/master/example/lib/main.dart)

```yaml
# pubspec.yaml

dependencies:
  flutter_image_utils: ^0.1.0+3
```

```dart
// E.g. lib/main.dart

import 'dart:typed_data';
import 'package:flutter_image_utils/flutter_image_utils.dart';

/* ... */

final img = AssetImage('assets/some_image.jpg');
final config = new ImageConfiguration();

final key = await img.obtainKey(config);
final data = await key.bundle.load(key.name);

final imgBytes = data.buffer.asUint8List();

// All other methods can be used same way
final cropped = await FlutterImageUtils.cropImage(
  imgBytes,
  x: 0,
  y: 0,
  width: 50,
  height: 50,
  quality: 85,
);

/* ... */

MemoryImage(Uint8List.fromList(cropped))
```


CHANGELOG.md
## 0.1.0+3

  * removed all `flutter_image_crop` and `FlutterImageCrop` occurences

## 0.1.0+2

* Fixed description
* Upgraded `image_picker` dev dependency
* Formatted `flutter_image_utils.dart`

## 0.1.0+1

* Fixed version in README

## 0.1.0

* Initial release. Implemented resize, resizeToMax, crop and rotate methods


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_image_utils_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_image_utils_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_image_utils/flutter_image_utils.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_image_utils');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterImageUtils.platformVersion, '42');
  });
}


example/lib/main.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_image_utils/flutter_image_utils.dart';

import 'package:image_picker/image_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Uint8List _imgBytes;
  Uint8List _croppedBytes;
  int _x = 0;
  int _y = 0;
  int _dw = 0;
  int _dh = 0;
  int _sw = 0;
  int _sh = 0;
  int _quality = 100;

  bool _isCropping = false;
  int _cropTime;

  Future<void> _pickImage(ImageSource source) async {
    final file = await ImagePicker.pickImage(source: source);

    if (file == null) {
      return;
    }

    final bytes = await file.readAsBytes();
    final uint8Bytes = Uint8List.fromList(bytes);

    final provider = MemoryImage(uint8Bytes);
    final completer = Completer<ImageInfo>();
    final stream = provider.resolve(ImageConfiguration());

    void listener(ImageInfo info, bool syncCall) => completer.complete(info);

    void errorListener(dynamic exception, StackTrace stackTrace) {
      completer.complete(null);
      FlutterError.reportError(new FlutterErrorDetails(
        context: 'image load failed ',
        library: 'flutter_image_utils',
        exception: exception,
        stack: stackTrace,
        silent: true,
      ));
    }

    stream.addListener(listener, onError: errorListener);
    final info = await completer.future;
    stream.removeListener(listener);

    setState(() {
      _cropTime = null;
      _imgBytes = uint8Bytes;
      _croppedBytes = uint8Bytes;
      _x = 0;
      _y = 0;
      _sw = info.image.width;
      _sh = info.image.height;
      _dw = _sw;
      _dh = _sh;
    });
  }

  Future<void> _crop() async {
    setState(() {
      _cropTime = null;
      _isCropping = true;
    });

    final stopwatch = Stopwatch()..start();
    final cropped = await FlutterImageUtils.cropImage(
      _imgBytes,
      x: _x,
      y: _y,
      width: _dw,
      height: _dh,
      quality: _quality,
    );

    setState(() {
      _cropTime = stopwatch.elapsedMilliseconds;
      _croppedBytes = cropped;
      _isCropping = false;
    });

    stopwatch.stop();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: const Text('flutter_image_utils'),
        ),
        body: SafeArea(
          child: SingleChildScrollView(
            child: Padding(
              padding: const EdgeInsets.all(16.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: <Widget>[
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    children: <Widget>[
                      Expanded(
                        child: FlatButton(
                          child: Text(
                            'camera',
                            style: TextStyle(color: Colors.white),
                          ),
                          color: Colors.blue,
                          shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(18.0),
                          ),
                          onPressed: () => _pickImage(ImageSource.camera),
                        ),
                      ),
                      Container(width: 16.0),
                      Expanded(
                        child: FlatButton(
                          child: Text(
                            'photo lib',
                            style: TextStyle(color: Colors.white),
                          ),
                          color: Colors.blue,
                          shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(18.0),
                          ),
                          onPressed: () => _pickImage(ImageSource.gallery),
                        ),
                      ),
                    ],
                  ),
                  Row(
                    children: <Widget>[
                      Expanded(
                        child: Text('x: $_x'),
                        flex: 1,
                      ),
                      Expanded(
                        child: Slider(
                          value: _x.toDouble(),
                          min: 0,
                          max: _sw.toDouble(),
                          divisions: _sw == 0 ? 1 : _sw,
                          onChanged: (v) => setState(() => _x = v.toInt()),
                        ),
                        flex: 3,
                      ),
                    ],
                  ),
                  Row(
                    children: <Widget>[
                      Expanded(
                        child: Text('y: $_y'),
                        flex: 1,
                      ),
                      Expanded(
                        child: Slider(
                          value: _y.toDouble(),
                          min: 0,
                          max: _sh.toDouble(),
                          divisions: _sh == 0 ? 1 : _sh,
                          onChanged: (v) => setState(() => _y = v.toInt()),
                        ),
                        flex: 3,
                      )
                    ],
                  ),
                  Row(
                    children: <Widget>[
                      Expanded(
                        child: Text('width: $_dw'),
                        flex: 1,
                      ),
                      Expanded(
                        child: Slider(
                          value: _dw.toDouble(),
                          min: 0,
                          max: _sw.toDouble(),
                          divisions: _sw == 0 ? 1 : _sw,
                          onChanged: (v) => setState(() => _dw = v.toInt()),
                        ),
                        flex: 3,
                      )
                    ],
                  ),
                  Row(
                    children: <Widget>[
                      Expanded(
                        child: Text('height: $_dh'),
                        flex: 1,
                      ),
                      Expanded(
                        child: Slider(
                          value: _dh.toDouble(),
                          min: 0,
                          max: _sh.toDouble(),
                          divisions: _sh == 0 ? 1 : _sh,
                          onChanged: (v) => setState(() => _dh = v.toInt()),
                        ),
                        flex: 3,
                      )
                    ],
                  ),
                  Row(
                    children: <Widget>[
                      Expanded(
                        child: Text('quality: $_quality'),
                        flex: 1,
                      ),
                      Expanded(
                        child: Slider(
                          value: _quality.toDouble(),
                          min: 0,
                          max: 100,
                          divisions: 100 == 0 ? 1 : 100,
                          onChanged: (v) => setState(() => _quality = v.toInt()),
                        ),
                        flex: 3,
                      )
                    ],
                  ),
                  FlatButton(
                    child: Text(
                      'crop',
                      style: TextStyle(color: Colors.white),
                    ),
                    color: Colors.blue,
                    disabledColor: Colors.grey,
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(18.0),
                    ),
                    onPressed: _isCropping || _imgBytes == null ? null : () => _crop(),
                  ),
                  _imgBytes != null
                      ? Container(
                          margin: const EdgeInsets.only(top: 16.0),
                          child: AspectRatio(
                            aspectRatio: _sw / _sh,
                            child: Container(
                              color: Colors.orange,
                              child: Image(
                                image: MemoryImage(_croppedBytes),
                              ),
                            ),
                          ),
                        )
                      : Container(),
                  _cropTime != null
                      ? Container(
                          child: Text('time elapsed: ${_cropTime}ms'),
                          margin: const EdgeInsets.only(top: 16.0),
                        )
                      : Container(),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}


ios/Classes/ImageUtils.swift
//
//  ImageUtils.swift
//  flutter_image_utils
//
//  Created by vanya elizarov on 20.06.2019.
//

import Foundation
import CoreGraphics

class ImageUtils {
    public static func crop(with data: Data, x: CGFloat, y: CGFloat, width: CGFloat, height: CGFloat, quality: Int) -> Data? {
        guard let img = UIImage(data: data) else { return nil }
        
        let rect = CGRect(x: x, y: y, width: width, height: height)
        
        guard let ref = img.cgImage else { return nil }
        guard let croppedRef = ref.cropping(to: rect) else { return nil }
        
        let croppedImg = UIImage(cgImage: croppedRef)
        let result = croppedImg.jpegData(compressionQuality: CGFloat(quality) / 100)
        
        return result
    }
    
    public static func resize(with data: Data, destWidth: CGFloat, destHeight: CGFloat, quality: Int) -> Data? {
        guard let img = UIImage(data: data) else { return nil }
        
        let widthRatio = destWidth  / img.size.width
        let heightRatio = destHeight / img.size.height
        
        var destSize: CGSize
        
        if widthRatio > heightRatio {
            destSize = CGSize(width: img.size.width * heightRatio, height: img.size.height * heightRatio)
        } else {
            destSize = CGSize(width: img.size.width * widthRatio, height: img.size.height * widthRatio)
        }
        
        let destRect = CGRect(x: 0, y: 0, width: destSize.width, height: destSize.height)
        
        UIGraphicsBeginImageContextWithOptions(destSize, false, 1.0)
        img.draw(in: destRect)
        let resizedImg = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        let result = resizedImg!.jpegData(compressionQuality: CGFloat(quality) / 100)
        
        return result
    }
    
    public static func resizeToMax(with data: Data, maxSize: CGFloat, quality: Int) -> Data? {
        guard let img = UIImage(data: data) else { return nil }
        
        var nextWidth = CGFloat(maxSize)
        var nextHeight = CGFloat(maxSize)
        
        if (img.size.width > img.size.height) {
            nextHeight = img.size.height * CGFloat(maxSize) / img.size.width;
        } else if (img.size.height > img.size.width) {
            nextWidth = img.size.width * CGFloat(maxSize) / img.size.height;
        }
        
        return ImageUtils.resize(with: data, destWidth: nextWidth, destHeight: nextHeight, quality: quality)
    }
    
    public static func rotate(with data: Data, angle: Int, quality: Int) -> Data? {
        guard let img = UIImage(data: data) else { return nil }
        
        let radians = CGFloat(angle) * .pi / 180
        var newSize = CGRect(origin: CGPoint.zero, size: img.size)
            .applying(CGAffineTransform(rotationAngle: radians))
            .size
        
        newSize.width = floor(newSize.width)
        newSize.height = floor(newSize.height)
        
        let destRect = CGRect(x: -img.size.width / 2, y: -img.size.height / 2, width: img.size.width, height: img.size.height)
        
        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        let context = UIGraphicsGetCurrentContext()!
        
        context.translateBy(x: newSize.width/2, y: newSize.height/2)
        context.rotate(by: CGFloat(radians))
        img.draw(in: destRect)
        
        let rotatedImg = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        let result = rotatedImg!.jpegData(compressionQuality: CGFloat(quality) / 100)
        
        return result
    }
}


ios/Classes/SwiftFlutterImageUtilsPlugin.swift
import Flutter
import UIKit

var serialQueue: DispatchQueue?

public class SwiftFlutterImageUtilsPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        serialQueue = DispatchQueue(label: "com.vanelizarov.flutter_image_utils.SerialQueue")
    
        let channel = FlutterMethodChannel(name: "flutter_image_utils", binaryMessenger: registrar.messenger())
        let instance = SwiftFlutterImageUtilsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        serialQueue?.sync {
            if call.method == "crop" {
                let args = call.arguments as! Dictionary<String, AnyObject>
                let bytes = args["bytes"] as! FlutterStandardTypedData
                let x = args["x"] as! CGFloat
                let y = args["y"] as! CGFloat
                let width = args["width"] as! CGFloat
                let height = args["height"] as! CGFloat
                let quality = args["quality"] as! Int
                
                let data = bytes.data
                
                guard let croppedData = ImageUtils.crop(
                    with: data,
                    x: x,
                    y: y,
                    width: width,
                    height: height,
                    quality: quality
                    ) else {
                        result(nil)
                        return
                }
                
                result(FlutterStandardTypedData(bytes: croppedData))
            } else if call.method == "resize" {
                let args = call.arguments as! Dictionary<String, AnyObject>
                let bytes = args["bytes"] as! FlutterStandardTypedData
                let destWidth = args["destWidth"] as! CGFloat
                let destHeight = args["destHeight"] as! CGFloat
                let quality = args["quality"] as! Int
                
                let data = bytes.data
                
                guard let resizedData = ImageUtils.resize(
                    with: data,
                    destWidth: destWidth,
                    destHeight: destHeight,
                    quality: quality
                    ) else {
                        result(nil)
                        return
                }
                
                result(FlutterStandardTypedData(bytes: resizedData))
            } else if call.method == "resizeToMax" {
                let args = call.arguments as! Dictionary<String, AnyObject>
                let bytes = args["bytes"] as! FlutterStandardTypedData
                let maxSize = args["maxSize"] as! CGFloat
                let quality = args["quality"] as! Int
                
                let data = bytes.data
                
                guard let resizedData = ImageUtils.resizeToMax(
                    with: data,
                    maxSize: maxSize,
                    quality: quality
                    ) else {
                        result(nil)
                        return
                }
                
                result(FlutterStandardTypedData(bytes: resizedData))
            } else if call.method == "rotate" {
                let args = call.arguments as! Dictionary<String, AnyObject>
                let bytes = args["bytes"] as! FlutterStandardTypedData
                let angle = args["angle"] as! Int
                let quality = args["quality"] as! Int
                
                let data = bytes.data
                
                guard let rotatedData = ImageUtils.rotate(
                    with: data,
                    angle: angle,
                    quality: quality
                    ) else {
                        result(nil)
                        return
                }
                
                result(FlutterStandardTypedData(bytes: rotatedData))
            } else {
                result(FlutterMethodNotImplemented)
            }
        }
    }
}


android/src/main/kotlin/com/vanelizarov/flutter_image_utils/CallHandler.kt
package com.vanelizarov.flutter_image_utils

import android.graphics.BitmapFactory
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executors

class CallHandler(var call: MethodCall, var result: MethodChannel.Result) {
    companion object {
        @JvmStatic
        private val executor = Executors.newFixedThreadPool(5)
    }

    fun handle() {
        executor.execute {
            when (call.method) {
                "crop" -> {
                    val bytes = call.argument<ByteArray>("bytes")!!
                    val x = call.argument<Int>("x")!!
                    val y = call.argument<Int>("y")!!
                    val width = call.argument<Int>("width")!!
                    val height = call.argument<Int>("height")!!
                    val quality = call.argument<Int>("quality")!!

                    try {

                        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.count())
                        val exifRotation = Exif.getRotationDegrees(bytes)

                        result.success(bitmap.crop(x, y, width, height, exifRotation).toByteArray(quality))
                    } catch (e: Exception) {
                        result.success(null)
                    }
                }

                "rotate" -> {
                    val bytes = call.argument<ByteArray>("bytes")!!
                    val angle = call.argument<Int>("angle")!!
                    val quality = call.argument<Int>("quality")!!

                    try {

                        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.count())
                        val exifRotation = Exif.getRotationDegrees(bytes)

                        result.success(bitmap.rotate(angle + exifRotation).toByteArray(quality))
                    } catch (e: Exception) {
                        result.success(null)
                    }
                }

                "resize" -> {
                    val bytes = call.argument<ByteArray>("bytes")!!
                    val destWidth = call.argument<Int>("destWidth")!!
                    val destHeight = call.argument<Int>("destHeight")!!
                    val quality = call.argument<Int>("quality")!!

                    try {

                        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.count())
                        val exifRotation = Exif.getRotationDegrees(bytes)

                        result.success(bitmap.resize(destWidth, destHeight, exifRotation).toByteArray(quality))
                    } catch (e: Exception) {
                        result.success(null)
                    }
                }

                "resizeToMax" -> {
                    val bytes = call.argument<ByteArray>("bytes")!!
                    val maxSize = call.argument<Int>("maxSize")!!
                    val quality = call.argument<Int>("quality")!!

                    try {

                        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.count())
                        val exifRotation = Exif.getRotationDegrees(bytes)

                        result.success(bitmap.resizeToMax(maxSize, exifRotation).toByteArray(quality))
                    } catch (e: Exception) {
                        result.success(null)
                    }
                }
            }
        }
    }
}

android/src/main/kotlin/com/vanelizarov/flutter_image_utils/BitmapExtension.kt
package com.vanelizarov.flutter_image_utils

import android.graphics.Bitmap
import android.graphics.Matrix
import java.io.ByteArrayOutputStream

fun Bitmap.crop(x: Int, y: Int, width: Int, height: Int, angle: Int): Bitmap {
    // compared to iOS cropping Android does not support rect overlapping
    // so we need to constrain dest width and height

    var sw = this.width
    var sh = this.height

    var dx = 0
    var dy = 0
    var dw = 0
    var dh = 0

    when (angle) {
        0 -> {
            dx = x
            dy = y
            dw = width
            dh = height
        }
        90 -> {
            dx = y
            dy = sh - x - width
            dw = height
            dh = width
        }
        180 -> {
            dx = sw - x - width
            dy = sh - y - height
            dw = width
            dh = height
        }
        270 -> {
            dx = sh - y - height
            dy = x
            dw = height
            dh = width
        }
    }

    dx = dx.coerceIn(0, sw)
    dy = dy.coerceIn(0, sh)
    dw = dw.coerceIn(0, sw - dx)
    dh = dh.coerceIn(0, sh - dy)

    val matrix = Matrix()
    matrix.preRotate(angle.toFloat())

    return  Bitmap.createBitmap(this, dx, dy, dw, dh, matrix, true)
}

fun Bitmap.rotate(angle: Int): Bitmap {
    return if (angle % 360 != 0) {
        val matrix = Matrix()
        matrix.setRotate(angle.toFloat())
        Bitmap.createBitmap(this, 0, 0, width, height, matrix, false)
    } else {
        this
    }
}

fun Bitmap.resize(destWidth: Int, destHeight: Int, angle: Int): Bitmap {
    return Bitmap.createScaledBitmap(this, destWidth, destHeight, true)
            .rotate(angle)
}

fun Bitmap.resizeToMax(maxSize: Int, angle: Int): Bitmap {

    var nextWidth = maxSize
    var nextHeight = maxSize

    if (this.width > this.height) {
        nextHeight = this.height * maxSize / this.width
    } else if (this.height > this.width) {
        nextWidth = this.width * maxSize / this.height
    }

    return Bitmap.createScaledBitmap(this, nextWidth, nextHeight, true)
            .rotate(angle)
}

fun Bitmap.toByteArray(quality: Int): ByteArray {
    val stream = ByteArrayOutputStream()
    this.compress(Bitmap.CompressFormat.JPEG, quality, stream)
    return stream.toByteArray()
}




























lib/flutter_image_utils.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/services.dart';

class FlutterImageUtils {
  static const MethodChannel _channel = const MethodChannel('flutter_image_utils');

  /// Returns `Future<Uint8List>` that contains bytes of cropped image.
  ///
  /// `x` and `y` params control the point from which the cropping
  /// will be performed.
  /// Must not be null or negative.
  ///
  /// `width` and `height` params control the width and the height
  /// of the cropping rect. They can be more than width and height
  /// of the original image, in this case platfrom code will constrain them
  /// to maximum possible.
  /// Must not be null or negative.
  ///
  /// `quality` param controls the quality of the output image. Must be
  /// positive number between `0` and `100`.
  ///
  /// ```dart
  /// import 'dart:typed_data';
  /// /* ... */
  /// final cropped = await FlutterImageUtils.cropImage(
  ///   imgBytes,
  ///   x: 0,
  ///   y: 0,
  ///   width: 50,
  ///   height: 50,
  ///   quality: 85,
  /// );
  /// /* ... */
  /// MemoryImage(cropped)
  /// ```
  static Future<Uint8List> cropImage(
    List<int> image, {
    int x,
    int y,
    int width,
    int height,
    int quality = 95,
  }) async {
    assert(image != null, 'image must not be null');

    assert(x != null, 'x must not be null');
    assert(y != null, 'y must not be null');
    assert(width != null, 'width must not be null');
    assert(height != null, 'height must not be null');
    assert(quality != null, 'quality must not be null');

    assert(x >= 0, 'x must be positive or equal to zero');
    assert(y >= 0, 'y must be positive or equal to zero');
    assert(width >= 0, 'width must be positive or equal to zero');
    assert(height >= 0, 'height must be positive or equal to zero');
    assert(quality >= 0, 'quality must be positive or equal to zero');

    if (image == null) {
      return Uint8List(0);
    }

    if (image.isEmpty) {
      return Uint8List(0);
    }

    final result = await _channel.invokeMethod('crop', <String, dynamic>{
      'bytes': Uint8List.fromList(image),
      'x': x,
      'y': y,
      'width': width,
      'height': height,
      'quality': quality,
    });

    if (result == null || result.isEmpty) {
      return Uint8List(0);
    }

    return _convertDynamic(result);
  }

  /// Returns `Future<Uint8List>` that contains bytes of proportionally scaled image.
  ///
  /// `maxSize` param controls max size to which image will be scaled
  /// E.g. if the image has size 1400x700 pixels and `maxSize` is set to 1200
  /// the size of the resulting image will be 1200x600
  /// Must not be null or negative.
  ///
  /// `quality` param controls the quality of the output image. Must be
  /// positive number between `0` and `100`.
  ///
  /// ```dart
  /// import 'dart:typed_data';
  /// /* ... */
  /// final sized = await FlutterImageUtils.resizeImageToMax(
  ///   imgBytes,
  ///   maxSize: 1200,
  ///   quality: 85,
  /// );
  /// /* ... */
  /// MemoryImage(sized)
  /// ```
  static Future<Uint8List> resizeImageToMax(
    List<int> image, {
    int maxSize,
    int quality = 95,
  }) async {
    assert(image != null, 'image must not be null');

    assert(maxSize != null, 'maxSize must not be null');
    assert(quality != null, 'quality must not be null');

    assert(maxSize >= 0, 'maxSize must be positive or equal to zero');
    assert(quality >= 0, 'quality must be positive or equal to zero');

    if (image == null) {
      return Uint8List(0);
    }

    if (image.isEmpty) {
      return Uint8List(0);
    }

    final result = await _channel.invokeMethod('resizeToMax', <String, dynamic>{
      'bytes': Uint8List.fromList(image),
      'maxSize': maxSize,
      'quality': quality,
    });

    if (result == null || result.isEmpty) {
      return Uint8List(0);
    }

    return _convertDynamic(result);
  }

  /// Returns `Future<Uint8List>` that contains bytes of resized image.
  ///
  /// `destWidth` and `destHeight` params control dimensions to which image will be scaled
  /// Must not be null or negative.
  ///
  /// `quality` param controls the quality of the output image. Must be
  /// positive number between `0` and `100`.
  ///
  /// ```dart
  /// import 'dart:typed_data';
  /// /* ... */
  /// final sized = await FlutterImageUtils.resizeImage(
  ///   imgBytes,
  ///   destWidth: 1200,
  ///   destWidth: 600,
  ///   quality: 85,
  /// );
  /// /* ... */
  /// MemoryImage(sized)
  /// ```
  static Future<Uint8List> resizeImage(
    List<int> image, {
    int destWidth,
    int destHeight,
    int quality = 95,
  }) async {
    assert(image != null, 'image must not be null');

    assert(destWidth != null, 'destWidth must not be null');
    assert(destHeight != null, 'destHeight must not be null');
    assert(quality != null, 'quality must not be null');

    assert(destWidth >= 0, 'destWidth must be positive or equal to zero');
    assert(destHeight >= 0, 'destHeight must be positive or equal to zero');
    assert(quality >= 0, 'quality must be positive or equal to zero');

    if (image == null) {
      return Uint8List(0);
    }

    if (image.isEmpty) {
      return Uint8List(0);
    }

    final result = await _channel.invokeMethod('resize', <String, dynamic>{
      'bytes': Uint8List.fromList(image),
      'destWidth': destWidth,
      'destHeight': destHeight,
      'quality': quality,
    });

    if (result == null || result.isEmpty) {
      return Uint8List(0);
    }

    return _convertDynamic(result);
  }

  /// Returns `Future<Uint8List>` that contains bytes of rotated image.
  ///
  /// `angle` param controls angle by which image will be rotated
  /// Must not be null or negative.
  ///
  /// `quality` param controls the quality of the output image. Must be
  /// positive number between `0` and `100`.
  ///
  /// ```dart
  /// import 'dart:typed_data';
  /// /* ... */
  /// final rotated = await FlutterImageUtils.rotateImage(
  ///   imgBytes,
  ///   angle: 90,
  ///   quality: 85,
  /// );
  /// /* ... */
  /// MemoryImage(rotated)
  /// ```
  static Future<Uint8List> rotateImage(
    List<int> image, {
    int angle,
    int quality = 95,
  }) async {
    assert(image != null, 'image must not be null');

    assert(angle != null, 'angle must not be null');
    assert(quality != null, 'quality must not be null');

    assert(quality >= 0, 'quality must be positive or equal to zero');

    if (image == null) {
      return Uint8List(0);
    }

    if (image.isEmpty) {
      return Uint8List(0);
    }

    final result = await _channel.invokeMethod('rotate', <String, dynamic>{
      'bytes': Uint8List.fromList(image),
      'angle': angle,
      'quality': quality,
    });

    if (result == null || result.isEmpty) {
      return Uint8List(0);
    }

    return _convertDynamic(result);
  }

  static Uint8List _convertDynamic(List<dynamic> list) {
    if (list == null || list.isEmpty) {
      return Uint8List(0);
    }

    return Uint8List.fromList(
        list.where((item) => item is int).map((item) => item as int).toList());
  }
}
